scripts/util/doc-builder.js

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env node
+
+var path = require('path');
+var fs = require('fs');
+var yuidoc = require('./yuidoc');
+var NodeGit = require('nodegit');
+var util = require('util');
+var fsExtra = require('fs-extra');
+
+function DocBuilder(options) {
+  this.options = util._extend({
+    masterBranchName: 'master',
+    docsBranchName: 'gh-pages',
+    docsDirName: 'docs',
+    versionsDirName: 'versions',
+    srcDirName: 'src',
+    rmSrcDir: false,
+    commitDocs: false,
+    checkoutStrategy: ( NodeGit.Checkout.STRATEGY.FORCE + NodeGit.Checkout.STRATEGY.DONT_WRITE_INDEX ),
+  }, options);
+
+  if (!this.options.themeDir) {
+    this.options.themeDir = path.join(this.options.docsDirName, 'yuidoc', 'theme');
+  }
+
+  this.options.apiDirPath = path.join(this.options.docsDirName, this.options.versionsDirName);
+}
+
+DocBuilder.prototype.ensureDocsBranchExists = function () {
+  console.info('info:', 'Ensuring \'' + this.options.docsBranchName + '\' branch exists');
+  return NodeGit.Repository.open('.').then(repo => {
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(references => {
+      var docsBranchReference;
+      var regex = new RegExp('/' + this.options.docsBranchName + '$'); //ensure reference name ends with the provided branch name
+
+      for (key in references) {
+        if (references[key].name().match(regex)) {
+          docsBranchReference = references[key];
+          if (!references[key].isRemote()) {
+            break;
+          }
+        }
+      };
```

Comment:
This loop is a little tricky to follow, because it's doing two things (making sure at least one reference matches the regex, checking if a reference is remote). It might be a bit cleaner to split this up into two pieces:

```javascript
docsBranchReferences = references.filter(function(key) {
  return references[key].name().match(regex);
});

if (!docsBranchReferences.length) {
  throw new Error(this.options.docsBranchName + ' could not be found');
}

remoteDocsBranchReferences = docsBranchReferences.filter(function(reference) {
  return reference.isRemote();
});
docsBranchReference = remoteDocsBranchReferences[0]

if (docsBranchReference) {
  // lines 50–57
}
```

Looking at that now, I'm not sure it's any easier to follow, but I like suggesting functional alternatives when I see 'em, even if it's just an interesting exercise for me to write out :)